### PR TITLE
DDF-3372 Update WebDav to not get file when deleting

### PR DESCRIPTION
**Pull request by peterhuffer**
#### What does this PR do?
- Update WebDav CDM to no longer attempt to fetch a file when a delete event is received
- Update FileSystemStorageProvider to catch IOException when failing to open inputStream to external reference

#### Who is reviewing it? 
@dcruver @brendan-hofmann @clockard @emmberk @AzGoalie 

#### Select relevant component teams: 
https://github.com/orgs/codice/teams
@codice/io 

#### Choose 2 committers to review/merge the PR. 
@lessarderic
@pklinef

#### How should this be tested? (List steps with links to updated documentation)
Follow instructions in description of https://codice.atlassian.net/browse/DDF-3372, except verify the exception does not occur.

Also verify that there is no Camel NPE warning when deleting a monitored item.

#### Any background context you want to provide?
When monitoring a WebDav server In Place with the CDM, the remote file would attempt to be downloaded even though it had already been deleted.

#### What are the relevant tickets?
[DDF-3372](https://codice.atlassian.net/browse/DDF-3372)

#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests
#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/DurableWebDavFileConsumer.java [hunk 1]
```diff
@@ -130,7 +130,11 @@ public void onDirectoryDelete(DavEntry entry) {
 
     @Override
     public void onFileDelete(DavEntry entry) {
-      processExchange(getExchange(entry, StandardWatchEventKinds.ENTRY_DELETE));
+      processExchange(
+          getExchange(
+              new File(entry.getLocation()),
```
Comment (brendan-hofmann) [inline catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/DurableWebDavFileConsumer.java]: ❗️ This change doesn't make sense to me. `getLocation()` will return a remote URI, which will never be resolvable as a `File`, plus it's been deleted anyways. I think you are just trying to bypass [this line](https://github.com/peterhuffer/ddf/blob/3b434dc20a5de287a5a4b7461558b139491debc2/catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/DurableWebDavFileConsumer.java#L85) but I don't think this is the right way to do that. There's already a null check [here](https://github.com/peterhuffer/ddf/blob/3b434dc20a5de287a5a4b7461558b139491debc2/catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/DurableWebDavFileConsumer.java#L85), so I think you are better off just passing null.

**Comment by clockard** [comment]
Once this pr is merged let's make sure we port this to the 2.11.x branch

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/DurableWebDavFileConsumer.java (hunk 1 above)
Comment (peterhuffer) [reply catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/DurableWebDavFileConsumer.java]: Yes, I was trying to pass that before, but I was getting an NPE down in Camel. I will see if I can find another way to get around that NPE.

Comment (brendan-hofmann) [reply catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/DurableWebDavFileConsumer.java]: I will also take a look on my end.

**Comment by peterhuffer** [comment]
Updated to pass null through the camel exchange. Also suppressed a camel warning about an NPE that it says it is ignoring.

File: distribution/ddf-common/src/main/resources/etc/org.ops4j.pax.logging.cfg [hunk 2]
```diff
@@ -105,6 +105,11 @@ log4j2.logger.org_apache_cxf_jaxrs_impl_WebApplicationExceptionMapper.level = ER
 log4j2.logger.org_apache_cxf_phase_PhaseInterceptorChain.name = org.apache.cxf.phase.PhaseInterceptorChain
 log4j2.logger.org_apache_cxf_phase_PhaseInterceptorChain.level = ERROR
 
+# camel
+# Supresses an NPE warning that is ignored by Camel
+log4j2.logger.org_apache_camel_impl_defaultunitofwork.level = ERROR
```
Comment (lessarderic) [inline distribution/ddf-common/src/main/resources/etc/org.ops4j.pax.logging.cfg]: ❓ How often is this exception happening?

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AbstractDurableFileConsumer.java [hunk 3]
```diff
@@ -118,8 +120,10 @@ public void onComplete(Exchange exchange) {
 
     @Override
     public void onFailure(Exchange exchange) {
-      INGEST_LOGGER.error(
-          "Delivery failed for {} event on {}", file, fileEvent.name(), exchange.getException());
+      if (INGEST_LOGGER.isErrorEnabled()) {
```
[the remark this answers is not in the document]
Comment (peterhuffer) [reply catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AbstractDurableFileConsumer.java]: This was a SonarLint suggestion.

File: distribution/ddf-common/src/main/resources/etc/org.ops4j.pax.logging.cfg (hunk 2 above)
Comment (peterhuffer) [reply distribution/ddf-common/src/main/resources/etc/org.ops4j.pax.logging.cfg]: It occurs every time a file being monitored is deleted.

File: catalog/core/catalog-core-camelcomponent/src/main/java/ddf/camel/component/catalog/content/ContentProducer.java [hunk 4]
```diff
@@ -79,16 +74,11 @@ public void process(Exchange exchange)
...
 
-    File ingestedFile = contentProducerDataAccessObject.getFileUsingRefKey(storeRefKey, in);
-
     WatchEvent.Kind<Path> eventType = contentProducerDataAccessObject.getEventType(storeRefKey, in);
 
-    if (ingestedFile == null) {
-      LOGGER.trace("EXITING: process - ingestedFile is NULL");
-      return;
-    }
+    File exchangeFile = contentProducerDataAccessObject.getFileUsingRefKey(storeRefKey, in);
```
[the remark this answers is not in the document]
Comment (peterhuffer) [reply catalog/core/catalog-core-camelcomponent/src/main/java/ddf/camel/component/catalog/content/ContentProducer.java]: Added a check for if the event is not delete, and the file is null to bail early. Also added unit tests.

File: catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AbstractDurableFileConsumer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lessarderic) [reply catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AbstractDurableFileConsumer.java]: Wish we could turn those off :-(

**Comment by peterhuffer** [comment]
build now

File: catalog/core/catalog-core-camelcomponent/src/main/java/ddf/camel/component/catalog/content/ContentProducerDataAccessObject.java [hunk 5]
```diff
@@ -137,10 +148,17 @@ public void createContentItem(
       throws SourceUnavailableException, IngestException {
     LOGGER.debug("Creating content item.");
 
+    if (!eventType.equals(ENTRY_DELETE) && ingestedFile == null) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(
+            "Ingested File was null with eventType [{}]. Doing nothing.", eventType.name());
```
Comment (ahoffer) [inline catalog/core/catalog-core-camelcomponent/src/main/java/ddf/camel/component/catalog/content/ContentProducerDataAccessObject.java]: Is `eventType.name()` an expensive call? Is that why there is an `if` guard `isDebugEnabled`?

File: catalog/core/catalog-core-camelcomponent/src/test/java/ddf/camel/component/catalog/content/ContentProducerDataAccessObjectTest.java [hunk 6]
```diff
@@ -88,6 +94,77 @@ public void testGetFileUsingRefKey() throws Exception {
...
+    doReturn(mockCatalogFramework).when(mockComponent).getCatalogFramework();
+
+    ContentEndpoint mockEndpoint = mock(ContentEndpoint.class);
+    doReturn(mockComponent).when(mockEndpoint).getComponent();
+
+    contentProducerDataAccessObject.createContentItem(
+        mockFileSystemPersistenceProvider, mockEndpoint, null, ENTRY_DELETE, "", headers);
+
+    verify(mockCatalogFramework, times(1)).delete(any(DeleteRequest.class));
+  }
```
Comment (ahoffer) [inline catalog/core/catalog-core-camelcomponent/src/test/java/ddf/camel/component/catalog/content/ContentProducerDataAccessObjectTest.java]: Nice use of Mockito in all your tests. They are really clean.

**Comment by peterhuffer** [comment]
build now

File: catalog/core/catalog-core-camelcomponent/src/main/java/ddf/camel/component/catalog/content/ContentProducerDataAccessObject.java (hunk 5 above)
Comment (peterhuffer) [reply catalog/core/catalog-core-camelcomponent/src/main/java/ddf/camel/component/catalog/content/ContentProducerDataAccessObject.java]: SonarLint gives warning on conditionally calling log statements. I didn't feel strongly either way, but wanted to be consistent if we are incorporating SonarLint into the development process.

**Comment by peterhuffer** [comment]
build now

**Comment by peterhuffer** [comment]
build now

**Comment by peterhuffer** [comment]
build now

**Comment by AzGoalie** [comment]
Verified on Windows 👍

**Pull request merged**